### PR TITLE
Add: replace Qwen3 decode paged attention with PyPTO

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

- replace the handwritten Qwen3-14B decode Page Attention CCE kernel with native PyPTO Phase 0, RoPE, and ragged GQA attention
- make PyPTO the sole production backend while preserving the serving ABI and arbitrary public-batch windowing
- preserve `PRE_LAUNCH=2` at the 512-token stack boundary with a three-slot transfer ring, so QK/softmax stays two complete stacks ahead of PV
- implement the final PV path as four internal 128-token pages per stack: prefetch every valid V page into four independent L1 `Mat` slots before waiting for `SOFTMAX_READY`, alternate P pages through two L1 buffers, and combine `P0@V0 + ... + P3@V3` in one FP32 accumulator using one `matmul` plus up to three `matmul_acc` operations
- close the remaining gap in the earlier 512-stack PyPTO path by restoring two CCE behaviors it did not explicitly preserve: move the complete valid V-stack load out of the post-`SOFTMAX_READY` critical path, and use an explicit two-slot L1 P ping-pong (`P0/P2` in slot 0, `P1/P3` in slot 1)
- keep one-, two-, and three-page final stacks branch-local, so missing pages issue no block-table/K/V/P loads, matmuls, or cross-core events while retaining exactly one QK/softmax/PV handoff per valid 512-token stack
- reduce the final B16/S3584 component effective mean from 308.034 us to 289.996 us (-5.86%) and the single-layer ragged decode mean from 802.9 us to 787.1 us (-1.97%) on A2/A3 device 8
- update component, full-decode, serving, contract, and documentation coverage

## Validation

- static: `py_compile`, Ruff, `git diff --check`, and all pre-commit hooks passed
- CPU: 8 Page Attention model tests and 31 Qwen3 contract tests passed
- A2/A3 simulator: the 7-case PR matrix compiled successfully across 5 signatures
- NPU component: the 7-case PR matrix passed, including one-, two-, and three-stack boundaries
- NPU integration: one-layer B17 decode matched 17/17 argmax and sampled tokens; logits were 100% within tolerance
- current PR head: all GitHub Actions checks passed, including generic A2/A3 Page Attention execution and Qwen3 serving accuracy
- repeated profiling: 10-layer and 40-layer production decode each completed 10 level-4 timing captures on even A2/A3 device 8; every 10-layer capture contained 10 PA tasks/720 worker slices, every 40-layer capture contained 40 PA tasks/2,880 worker slices, and no records were dropped

### Final 512-token stack optimization

The final commit keeps cross-core communication at 512-token granularity while retaining 128-token Cube work internally:

1. QK processes the four physical pages with a two-stage PyPTO pipeline and publishes one 512-token score tile.
2. AIV performs one 512-token online-softmax update and publishes P once for the stack.
3. AIC loads all valid V pages into four distinct L1 `Mat` slots before the softmax-ready wait, overlapping V-cache traffic with AIV softmax.
4. After the wait, P ping-pongs through two L1 slots while one FP32 `Acc` tile receives one `matmul` and up to three `matmul_acc` updates.
5. The final PV tile is published once. A 1-3 page tail uses branch-local fixed paths and never computes a virtual page.

This preserves the original three handoffs per 512-token stack and the two-stack prelaunch schedule; only the internal PyPTO data movement and Cube accumulation are changed. The focused A/B used the same A2/A3 device 8, deterministic B16 inputs, 5 warmups, 100 resident dispatches, and an in-run golden.

The earlier 512-stack implementation was slower for two concrete reasons:

| Missing behavior in the earlier PyPTO path | Final change | Handwritten CCE mapping |
|---|---|---|
| V was loaded page-by-page after `SOFTMAX_READY`, placing Value Cache traffic on the PV critical path | Load every valid V page into a distinct L1 `Mat` slot before the single stack-level wait | `BlockMmadPV` gathers the valid V stack into L1 before waiting for P/softmax readiness |
| P and V were coupled inside a generic dynamic page loop, without an explicit CCE-style P buffer lifetime | Alternate `P0/P1/P2/P3` through two named L1 `Mat` slots (`0/1/0/1`) while accumulating into one FP32 tile | CCE uses P ping-pong buffers while four 128-token MMADs accumulate in one L0C |

The committed implementation first makes all four V pages resident, then performs the only stack-level wait ([full source](https://github.com/hw-native-sys/pypto-lib/blob/1406fc635cff7912a3917ae41d6ec74fa3d5f037/models/qwen3_14b/paged_attention_pypto.py#L400-L448)):

```python
# Abridged: repeated shape and address arguments are omitted.
v0: pl.Tile[..., pl.MemRef("pv_v_l1", slots=STACK_PAGES)[0], pl.Mem.Mat] = pl.load(...)
v1: pl.Tile[..., pl.MemRef("pv_v_l1", slots=STACK_PAGES)[1], pl.Mem.Mat] = pl.load(...)
v2: pl.Tile[..., pl.MemRef("pv_v_l1", slots=STACK_PAGES)[2], pl.Mem.Mat] = pl.load(...)
v3: pl.Tile[..., pl.MemRef("pv_v_l1", slots=STACK_PAGES)[3], pl.Mem.Mat] = pl.load(...)
pl.system.sync_wait(SOFTMAX_READY_EVENT, pipe=pl.PipeType.MTE2, core_type="aic")
```

After the wait, P uses the explicit `0/1/0/1` L1 ping-pong and the same FP32 accumulator ([full source](https://github.com/hw-native-sys/pypto-lib/blob/1406fc635cff7912a3917ae41d6ec74fa3d5f037/models/qwen3_14b/paged_attention_pypto.py#L449-L517)):

```python
probability0: pl.Tile[..., pl.MemRef("pv_p_l1", slots=2)[0], pl.Mem.Mat] = pl.load(...)
pv0 = pl.matmul(probability0, v0, out_dtype=pl.FP32)
probability1: pl.Tile[..., pl.MemRef("pv_p_l1", slots=2)[1], pl.Mem.Mat] = pl.load(...)
pv1 = pl.matmul_acc(pv0, probability1, v1)
probability2: pl.Tile[..., pl.MemRef("pv_p_l1", slots=2)[0], pl.Mem.Mat] = pl.load(...)
pv2 = pl.matmul_acc(pv1, probability2, v2)
probability3: pl.Tile[..., pl.MemRef("pv_p_l1", slots=2)[1], pl.Mem.Mat] = pl.load(...)
pv3 = pl.matmul_acc(pv2, probability3, v3)
```

| Scope | Before final V prefetch (us) | Current `1406fc6` (us) | Change |
|---|---:|---:|---:|
| B16 / S3584 fused Phase-0 + PA effective mean | 308.034 | 289.996 | -5.86% |
| B16 ragged single-layer decode effective mean | 802.9 | 787.1 | -1.97% |

### 10-layer and 40-layer comparison with local handwritten CCE main

Both implementations used A2/A3 **device 8**, seed 1234, the same explicit B16 ragged sequence lengths (`2864,724,1319,3126,3277,3445,665,2042,3728,2738,280,1258,3227,31,3756,1183`), and 10 clean level-4 timing captures per layer count. Dependency-graph capture timing was discarded. The common toolchain was PyPTO `d2cf77c`, Simpler `1f27a157`, PTO-ISA `f51c92f`, and PTOAS 0.57.

The PR row is `1406fc6`. The handwritten baseline is the requested local `~/workspace/pypto-lib` `main` checkout at `384553a`, plus its two pre-existing `Tensor` to `ChipTensor` ABI compatibility edits in the CCE kernel/tiler; those edits do not alter the Page Attention schedule. It is therefore labeled as the local-main baseline rather than pristine current `origin/main`.

`PA worker Exec mean` averages setup-free kernel execution across every PA worker slice and all 10 captures. `PA layer-critical mean` first takes the slowest AIC/AIV worker in each layer, then averages layers and captures. `Run-global PA max mean` takes the single slowest PA worker across the complete graph in each capture, then averages the 10 maxima. `AICore E2E mean` spans earliest scheduler dispatch to latest AICore finish for the complete 10-layer/40-layer graph.

| Backend | Layers | Captures | PA worker Exec mean (us) | PA layer-critical mean (us) | Run-global PA max mean (us) | AICore E2E mean (us) | E2E vs local main |
|---|---:|---:|---:|---:|---:|---:|---:|
| Handwritten CCE local main (`384553a` + ABI compatibility) | 10 | 10 | 170.663 | 201.787 | 209.424 | 9,537.710 | — |
| PyPTO PR (`1406fc6`) | 10 | 10 | 164.567 | 199.014 | 215.264 | 9,492.522 | -45.188 us (-0.47%) |
| Handwritten CCE local main (`384553a` + ABI compatibility) | 40 | 10 | 170.285 | 201.263 | 231.146 | 33,428.670 | — |
| PyPTO PR (`1406fc6`) | 40 | 10 | 165.131 | 199.910 | 216.552 | 33,227.642 | -201.028 us (-0.60%) |

The PR lowers PA worker mean by 3.57% at 10 layers and 3.03% at 40 layers; the layer-critical mean improves by 1.37% and 0.67%, respectively. The run-global maximum is an extreme-order statistic and was visibly noisier for the CCE baseline (about 3.9% CV versus at most 0.74% for the PR), so AICore E2E is the primary wall-time result. Its CV stayed below 0.26% for all four configurations.

These level-4 captures are attribution runs: `kernel-duration-us` excludes runtime local setup but includes the fused kernel body, and profiling mode intentionally skips host-reference replay. Correctness is covered separately by the device component matrix and decode validation listed above.

### Historical PRE_LAUNCH scheduling A/B

This earlier same-device A/B compared lockstep `f7a3aa9` with the prefetch head `bd1d1a2`. The fused Phase-0 + Page Attention B16/S3584 case used one timed process per variant, 10 warmups, and 100 resident dispatches. Values are the mean of `[RUN] effective_us`; each variant independently generated the same deterministic Torch oracle and both passed the `key_cache`, `value_cache`, and `out` checks.

Toolchain: PyPTO `d2cf77c`, Simpler `1f27a157`, PTO-ISA `f51c92f`, PTOAS 0.57.

| Schedule | Effective mean (us) | Effective median (us) | Change vs lockstep |
|---|---:|---:|---:|
| Lockstep PyPTO (`f7a3aa9`) | 408.916 | 408.740 | — |
| `PRE_LAUNCH=2` PyPTO (`bd1d1a2`) | 345.030 | 345.090 | -15.62% |

That PRE_LAUNCH change reduced mean effective wall time by 63.886 us (15.62%) relative to the lockstep baseline. Device-wall mean also decreased from 427.654 us to 363.538 us (-14.99%). The measured source also included the valid-slot clamp; all benchmark slots were already valid, so it did not change the tested workload.

### Historical backend A/B before Phase-0 fusion and PRE_LAUNCH

These earlier migration measurements predate the final fused Phase-0 scope and `PRE_LAUNCH=2`; they are retained as history and are not directly comparable with the current-head table above. Each backend used 5 fresh processes, with 5 warmups and 100 resident dispatches per process. Values are aggregate means.

| Scope | Device | Handwritten CCE (us) | Early PyPTO (us) | Delta |
|---|---:|---:|---:|---:|
| L40 / B16 / S3584 full decode | 2 | 37,327.713536 | 39,034.081424 | +4.571316% |
| L1 / B16 / S3584 single layer | 6 | 2,471.014718 | 2,504.192982 | +1.342698% |
| B16 / S3584 Page Attention only | 4 | 301.724152 | 315.016116 | +4.405336% |

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -342,6 +352,148 @@ jobs:
...
+        shell: bash
+        run: |
+          source activate.sh
+          PYTHONPATH="$PWD/models/qwen3_14b:$PWD" \
+            python -m pytest \
+              models/qwen3_14b/test_decode_pa_perf.py \
+              models/qwen3_14b/test_paged_attention_pypto.py \
+              models/qwen3_14b/test_decode_pa_ab.py -q
+          PYTHONPATH="$PWD" \
+            python -m pytest tests/contract/test_qwen3_14b_contract.py -q
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply .github/workflows/ci.yml]: Fixed in 2d497f0: qwen-pa-contract, qwen-pa-a2a3sim, qwen-pa-device, and nightly qwen-pa-full now each declare least-privilege `permissions: contents: read`.

File: models/qwen3_14b/decode_fwd.py [hunk 2]
```diff
@@ -1599,50 +1872,124 @@ def _build_specs(inputs: dict) -> list:
...
+        if args.max_seq:
+            parser.error("--seq-lens and --max-seq are mutually exclusive")
+        try:
+            requested_seq_lens = [int(value) for value in args.seq_lens.split(",")]
+        except ValueError:
+            parser.error("--seq-lens must be a comma-separated list of integers")
+        if len(requested_seq_lens) != BATCH_PAD or any(
+            not 1 <= value <= MAX_SEQ for value in requested_seq_lens
+        ):
+            parser.error(f"--seq-lens must contain {BATCH_PAD} values in [1, {MAX_SEQ}]")
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply models/qwen3_14b/decode_fwd.py]: Fixed in 2d497f0: `--seq-lens` now validates against `args.batch` for `--validate-fwd` and remains BATCH_PAD-sized for the static single-layer driver. The device lane now exercises the B17 case.

File: models/qwen3_14b/paged_attention_pypto.py [hunk 3]
```diff
@@ -0,0 +1,816 @@
...
+                )
+                s0_scaled = pl.tile.muls(s0, SCALE)
+                valid_cols0 = pl.min(BLOCK_SIZE, seq_len)
+                s0_valid = pl.set_validshape(s0_scaled, lane_rows, valid_cols0)
+                s0_filled = pl.fillpad(s0_valid, pad_value=pl.PadValue.min)
+                s0_masked = pl.set_validshape(s0_filled, lane_rows, BLOCK_SIZE)
+                zrow = pl.tile.muls(pl.row_max(s0_masked, tmp), 0.0)
+                one = pl.tile.adds(zrow, 1.0)
+                m = pl.tile.adds(zrow, -3.0e38)
+                l = zrow
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply models/qwen3_14b/paged_attention_pypto.py]: Fixed in 2d497f0: renamed the online-softmax denominator accumulator and every use to `l_sum`; Ruff passes.

File: models/qwen3_14b/test_decode_pa_perf.py [hunk 4]
```diff
@@ -0,0 +1,1329 @@
...
+            f"PTO2_MANUAL_MAX_SEQ={configured_max_seq!r} disagrees with "
+            f"fixture max_seq={fixture_max_seq}; use a fresh process"
+        )
+    os.environ["PTO2_MANUAL_MAX_SEQ"] = fixture_max_seq
+    decode_module = importlib.import_module("decode_fwd")
+    if decode_module.MAX_SEQ != manifest["max_seq"]:
+        raise RuntimeError(
+            f"decode_fwd MAX_SEQ={decode_module.MAX_SEQ} does not match fixture {manifest['max_seq']}"
+        )
+    decode_module._FWD_NLAYERS = manifest["layers"]
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply models/qwen3_14b/test_decode_pa_perf.py]: Resolved by removing this obsolete migration-only benchmark in 2d497f0. Raw informational timing remains in the focused component driver and no longer mutates `decode_fwd._FWD_NLAYERS` through an imported benchmark harness.

**Comment by Little-oil** [comment]
修改前cce时长：
image

**Comment by Little-oil** [comment]
改后PRE_LAUNCH=2， STACK_TOKENS从512改成128，使用独立 K/V ping-pong MemRef
max慢8us，mean慢1us：
image

**Comment by Little-oil** [comment]
去掉奇数尾页虚算
快1us，基本没变：
image

**Comment by Little-oil** [comment]
V在等soft_ready之前一次性取完+P buffer固定P0/P2->buffer0, P1/P3->buffer1：
和cce的时间基本相同
image
